### ISSUE TITLE
Improve performance when you have 10,000 ocean depth cache objects

### ISSUE THREAD

**Issue by dizzy2003** [crest-ocean/crest#333]
Imagine thousands of coral objects each with a depth cache in the scene.

Not sure if a solution has allready been created for this issue. But last I looked each OceanDepthCache registered a quad to be rendered, there was no culling on these objects so they were all submitted as draw calls via a command buffer.

I actually have a local solution I implemented where I have a grid of 9 rendertargets around the main camera each render target is itself an OceanDepthCache. These render targets are populated only when the camera moves across a grid boundary. (and repositioned) They are rendered simply with a unity camera  that renders a single layer and therefore does its own culling. I need to do a bit more work on this system as right now it re renders all 9 targets as the camera crosses the boundary rather than shuffling the targets around and just updating the new ones. I am happy to provide this code back to the community if its desired. allthought its in our own coding standards)  I also added a shader depth thats like geometry depth but also uses a texture thats a alpha blob to blend out the depth at the edges.

An alternative solution would be to use culling groups for everything submitted via RegisterSeaFloorDepthInput so that only visible objects get rendererd (potentially re building the command buffer a lot as objects come in and out of view). I have used unity culling groups elsewhere and they are quite performant.

**Comment by huwb** [comment]
Interesting, I hadn't thought about the case where there are lots of small caches. Would it be possible to cover the 10000 coral objects with a small number of large depth caches?

It's a good point in any case they we don't cull any ocean inputs. Culling would also be helpful for our under water effect rewrite. I did look into it once in the past but didn't find it easy to configure the culling query. Maybe it's been simplified since I looked, or I missed a trick with how to set it up.

**Comment by dizzy2003** [comment]
in our case the coral is spread over an area 8kmx8km and are randomly
positioned each time you play.

**Comment by huwb** [comment]
Ahh ok.

I read up on culling groups and they're straightforward nowadays but seem to have a hard dependence on providing a camera. this is unfortunate as there is no explicit camera for rendering the LOD data. 

Maybe you meant use the 'main' camera to cull the inputs to the ocean. thats not how we generally do things with the ocean. there are persistent simulations (foam, wave) which ideally want consistent inputs that don't vary greatly when the view changes. this might work ok for some/many cases though.

An alternative would be to do the culling ourselves. this would test each input against each cache rect. this could potentially be on a burst job and would likely be quite cheap - but would require a late 2018 or 2019 version of unity.

If i'm understanding correctly, the last option sounds like the most general. Let me know what you think.

**Comment by dizzy2003** [comment]
Theres nothing to stop a fake camera being used that renders no layers and
sole purpose is to provide a cull frustum, which I would imagine would
actually be a big volume round the player. Maybe the camera will still cull
groups when its disabled so there will be no overhead of an active camera.

maybe to keep the sim happy, each lod would have a lower
_attenuationInShallows to the effect its almost non existant in the outer
lod.

you dont like my system of auto maintaining a rendered cache of quads round
the camera that contain ocean depth info for large numbers of individual
objects? Its effectively anothe layer built on top of the existing cached
depths system.

The upside of that is it only renders objects when the camera crosses a
boundary to a new grid cell. The downside I guess is memory for the 9
render targets, performance overhead is spikey but mostly 0 and it requires
a unity layer to let the unity camera do the culling. <allthough it could
use its own cullgroups and build command buffers to handle the rendering>

Am not sure what the overhead might at your end of the simulation with 9
large quads of OceanDepthCaches.

**Comment by huwb** [comment]
Good point - i could set up a set of cameras that are disabled which i move around to drive the culling for each lod. would be an interesting test.

Yeah i think your system is a good idea. If you'd be up for contributing the code back that would be valuable. the reason my mind went to culling is because im always thinking about the most general approaches when thinking about adding stuff to core, and culling would work for all ocean input types, and dynamically changing depth. however if there are a lot of depth cache objects, it will still do a lot of draw calls into the largest lod. so perhaps culling and your tiled system are orthogonal solutions (both could be effectively applied to solve different problems).

a last thought is that instancing the cached depth quads would likely help draw calls, but i dont know how that interacts with culling. maybe something to look into later down the line.

**Comment by dizzy2003** [comment]
yeah instancing would be a massive help cpu side at least.

**Comment by huwb** [comment]
Sorry there has been no movement on this. I havent forgotten this and i would like to hit it. I think its not just lots of depth caches, but even a single depth cache populates the water simulation data using multiple draw calls and could be done in a single compute dispatch.

I'm not sure when we'll get to this as our current resources only permit us to work on a small number of features, after support and etc. We're hoping to increase our bandwidth through sponsorship and i wanted to mention that once here. https://github.com/sponsors/wave-harmonic
